Errors using example run files with latest version of DifferentialEquations.jl

Newer versions of DifferentialEquations.jl (and its dependency RecursiveArrayTools.jl) changed the way the solution can be accessed as an array, causing the example run files to error out. Indeed, when running the dynamic files, result in the following error:

```bash
% julia run/run_dynamics.jl
Starting...
ERROR: LoadError: TaskFailedException

    nested task error: AssertionError: length(I) == ndims(A.u)
    Stacktrace:
   ....
```

Apparently, the way the solution was used as an array, although it worked fine before, it was never officially supported (see: https://github.com/SciML/DifferentialEquations.jl/issues/1021). The official way to access the solution from DifferentialEquations.jl is to do `Array(sol)'.

--- a/runs/run_dynamics.jl
+++ b/runs/run_dynamics.jl
@@ -46,7 +46,7 @@ Threads.@threads for i in 1:navg
                bfield(N, Δt, J, noise),
                bfield(N, Δt, J, noise)];
     sol = diffeqsolver(s0, tspan, J, bfields, matrix; JH=JH, saveat=saveat);
-    sols[i, :, :] = transpose(sol[:, :])
+    sols[i, :, :] = transpose(Array(sol)[:, :])
     next!(progress)
 end
 solavg = mean(sols, dims=1)[1, :, :];

--- a/runs/run_dynamics.jl
+++ b/runs/run_dynamics.jl
@@ -40,16 +40,16 @@ JH = Nchain(nspin, J0)
 
 println("Starting...")
 progress = Progress(navg);
-sols = zeros(navg, length(saveat), 3*nspin)
+sols = zeros(navg, 3*nspin, length(saveat))
 Threads.@threads for i in 1:navg
     bfields = [bfield(N, Δt, J, noise),
                bfield(N, Δt, J, noise),
                bfield(N, Δt, J, noise)];
     sol = diffeqsolver(s0, tspan, J, bfields, matrix; JH=JH, saveat=saveat);
-    sols[i, :, :] = transpose(Array(sol)[:, :])
+    sols[i, :, :] = Array(sol)
     next!(progress)
 end
-solavg = mean(sols, dims=1)[1, :, :];
+solavg = dropdims(mean(sols, dims=1), dims=1);
 
 ########################
 ########################
@@ -60,17 +60,17 @@ npzwrite("./dynamics.npz", Dict("t" => saveat,
 
 ### Save data CSV ###
 dataframe = DataFrame(t = saveat,
-                      Sx = solavg[:, 1],
-                      Sy = solavg[:, 2],
-                      Sz = solavg[:, 3])
+                      Sx = solavg[1, :],
+                      Sy = solavg[2, :],
+                      Sz = solavg[3, :])
 CSV.write("./dynamics.csv", dataframe)
 
 ### Plots ###
-plot(saveat, solavg[:, 1], xlabel="t", ylabel="S_x")
+plot(saveat, solavg[1, :], xlabel="t", ylabel="S_x")
 savefig("./sx.pdf")
 
-plot(saveat, solavg[:, 2], xlabel="t", ylabel="S_y")
+plot(saveat, solavg[2, :], xlabel="t", ylabel="S_y")
 savefig("./sy.pdf")
 
-plot(saveat, solavg[:, 3], xlabel="t", ylabel="S_z")
+plot(saveat, solavg[3, :], xlabel="t", ylabel="S_z")
 savefig("./sz.pdf")

--- a/runs/run_dynamics_ho.jl
+++ b/runs/run_dynamics_ho.jl
@@ -43,8 +43,8 @@ Threads.@threads for i in 1:navg
                bfield(N, Δt, J, noise),
                bfield(N, Δt, J, noise)];
     sol = diffeqsolver(x0, p0, tspan, J, bfields, matrix; saveat=saveat);
-    solx[i, :, :] = transpose(sol[1:3*nosc, :])
-    solp[i, :, :] = transpose(sol[1+3*nosc:6*nosc, :])
+    solx[i, :, :] = transpose(Array(sol)[1:3*nosc, :])
+    solp[i, :, :] = transpose(Array(sol)[1+3*nosc:6*nosc, :])
     next!(progress)
 end
 

--- a/runs/run_dynamics_ho.jl
+++ b/runs/run_dynamics_ho.jl
@@ -35,21 +35,21 @@ navg = 10
 println("Starting...")
 
 progress = Progress(navg);
-solx = zeros(navg, length(saveat), 3*nosc)
-solp = zeros(navg, length(saveat), 3*nosc)
+solx = zeros(navg, 3*nosc, length(saveat))
+solp = zeros(navg, 3*nosc, length(saveat))
 
 Threads.@threads for i in 1:navg
     bfields = [bfield(N, Δt, J, noise),
                bfield(N, Δt, J, noise),
                bfield(N, Δt, J, noise)];
     sol = diffeqsolver(x0, p0, tspan, J, bfields, matrix; saveat=saveat);
-    solx[i, :, :] = transpose(Array(sol)[1:3*nosc, :])
-    solp[i, :, :] = transpose(Array(sol)[1+3*nosc:6*nosc, :])
+    solx[i, :, :] = Array(sol)[1:3*nosc, :]
+    solp[i, :, :] = Array(sol)[1+3*nosc:6*nosc, :]
     next!(progress)
 end
 
-solavgx = mean(solx, dims=1)[1, :, :];
-solavgp = mean(solp, dims=1)[1, :, :];
+solavgx = dropdims(mean(solx, dims=1), dims=1);
+solavgp = dropdims(mean(solp, dims=1), dims=1);
 
 ########################
 ########################
@@ -61,29 +61,29 @@ npzwrite("./dynamics.npz", Dict("t" => saveat,
 
 ### Save data CSV ###
 dataframe = DataFrame(t = saveat,
-                      xx = solavgx[:, 1],
-                      xy = solavgx[:, 2],
-                      xz = solavgx[:, 3],
-                      px = solavgp[:, 1],
-                      py = solavgp[:, 2],
-                      pz = solavgp[:, 3])
+                      xx = solavgx[1, :],
+                      xy = solavgx[2, :],
+                      xz = solavgx[3, :],
+                      px = solavgp[1, :],
+                      py = solavgp[2, :],
+                      pz = solavgp[3, :])
 CSV.write("./dynamics.csv", dataframe)
 
 ### Plots ###
-plot(saveat, solavgx[:, 1], xlabel="t", ylabel="x_x")
+plot(saveat, solavgx[1, :], xlabel="t", ylabel="x_x")
 savefig("./xx.pdf")
 
-plot(saveat, solavgx[:, 2], xlabel="t", ylabel="x_y")
+plot(saveat, solavgx[2, :], xlabel="t", ylabel="x_y")
 savefig("./xy.pdf")
 
-plot(saveat, solavgx[:, 3], xlabel="t", ylabel="x_z")
+plot(saveat, solavgx[3, :], xlabel="t", ylabel="x_z")
 savefig("./xz.pdf")
 
-plot(saveat, solavgp[:, 1], xlabel="t", ylabel="p_x")
+plot(saveat, solavgp[1, :], xlabel="t", ylabel="p_x")
 savefig("./px.pdf")
 
-plot(saveat, solavgp[:, 2], xlabel="t", ylabel="p_y")
+plot(saveat, solavgp[2, :], xlabel="t", ylabel="p_y")
 savefig("./py.pdf")
 
-plot(saveat, solavgp[:, 3], xlabel="t", ylabel="p_z")
+plot(saveat, solavgp[3, :], xlabel="t", ylabel="p_z")
 savefig("./pz.pdf")

--- a/runs/run_example_paper.jl
+++ b/runs/run_example_paper.jl
@@ -42,7 +42,7 @@ for i in 1:ntraj # iterations through the number of trajectories
                      bfield(N, Δt, Jsd, noise)] # stochastic fields
     # diffeqsolver (below) solves the system for the single trajectory
     local sol = diffeqsolver(s0, tspan, Jsd, bfields, Cw; saveat=saveat)
-    sols[i,:,:] = sol[:,:] # store the trajectory into the matrix of solutions
+    sols[i,:,:] = Array(sol) # store the trajectory into the matrix of solutions
 end
 
 ### example plot ###

--- a/runs/run_steadystate.jl
+++ b/runs/run_steadystate.jl
@@ -49,7 +49,7 @@ for n in eachindex(T)
                    bfield(N, Δt, J, noise),
                    bfield(N, Δt, J, noise)];
         sol = diffeqsolver(s0, tspan, J, bfields, matrix; JH=JH, saveat=saveat);
-        s[i, :] = mean(sol, dims=2)
+        s[i, :] = mean(Array(sol), dims=2)
     end
     Sss[n, :] = mean(s, dims=1)
     next!(progress)

--- a/runs/run_steadystate_ho.jl
+++ b/runs/run_steadystate_ho.jl
@@ -24,7 +24,7 @@ T = 10 .^ LinRange(-3, 3, 24)
 
 navg = 10 # number of stochastic field realizations to average
 
-N = 1
+nosc = 1
 
 ########################
 ########################
@@ -32,13 +32,13 @@ N = 1
 println("Starting...")
 
 progress = Progress(length(T));
-xss = zeros(length(T), 3*N)
-pss = zeros(length(T), 3*N)
+xss = zeros(length(T), 3*nosc)
+pss = zeros(length(T), 3*nosc)
 
 for n in eachindex(T)
     noise = ClassicalNoise(T[n]);
-    x = zeros(navg, 3*N)
-    p = zeros(navg, 3*N)
+    x = zeros(navg, 3*nosc)
+    p = zeros(navg, 3*nosc)
     Threads.@threads for i in 1:navg
         x0 = [1., 0., 0.]
         p0 = [0., 0., 0.]
@@ -46,8 +46,8 @@ for n in eachindex(T)
                    bfield(N, Δt, J, noise),
                    bfield(N, Δt, J, noise)];
         sol = diffeqsolver(x0, p0, tspan, J, bfields, matrix; saveat=saveat);
-        x[i, :] = mean(sol[1:3*N, :].^2, dims=2)
-        p[i, :] = mean(sol[1+3*N:6*N, :].^2, dims=2)
+        x[i, :] = mean(sol[1:3*nosc, :].^2, dims=2)
+        p[i, :] = mean(sol[1+3*nosc:6*nosc, :].^2, dims=2)
     end
     xss[n, :] = mean(x, dims=1)
     pss[n, :] = mean(p, dims=1)

--- a/runs/run_steadystate_ho.jl
+++ b/runs/run_steadystate_ho.jl
@@ -46,8 +46,8 @@ for n in eachindex(T)
                    bfield(N, Δt, J, noise),
                    bfield(N, Δt, J, noise)];
         sol = diffeqsolver(x0, p0, tspan, J, bfields, matrix; saveat=saveat);
-        x[i, :] = mean(sol[1:3*nosc, :].^2, dims=2)
-        p[i, :] = mean(sol[1+3*nosc:6*nosc, :].^2, dims=2)
+        x[i, :] = mean(Array(sol)[1:3*nosc, :].^2, dims=2)
+        p[i, :] = mean(Array(sol)[1+3*nosc:6*nosc, :].^2, dims=2)
     end
     xss[n, :] = mean(x, dims=1)
     pss[n, :] = mean(p, dims=1)

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -67,7 +67,7 @@ using Test
             
                 s0 = normalize(rand(3))
                 sol = diffeqsolver(s0, tspan, J, bfields, Cw; saveat=saveat, alg=Tsit5(), atol=1e-5, rtol=1e-5);
-                sdynss = mean(sol, dims=2)
+                sdynss = mean(Array(sol), dims=2)
             
                 # analytic solution of T=0 cl steady-state
                 norm_eq_coeffs(ζ, θ0) = [-1, 4*ζ, 1 - 4*ζ^2, -4*ζ*sin(θ0)^2, 4*ζ^2*sin(θ0)^2]
@@ -104,7 +104,7 @@ using Test
         
             s0 = normalize(rand(3))
             sol = diffeqsolver(s0, tspan, J, bfields, Cw; saveat=saveat, alg=Tsit5(), atol=1e-5, rtol=1e-5);
-            sdynss = mean(sol, dims=2)
+            sdynss = mean(Array(sol), dims=2)
 
             @test isapprox(sdynss[:,end], [-0.53567, -0.11054, 0.81484], atol=1e-5)
         end

